src/workerd/api/messagechannel.c++

```diff
@@ -0,0 +1,317 @@
+      }
+      KJ_CASE_ONEOF(opts, PostMessageOptions) {
+        KJ_IF_SOME(list, opts.transfer) {
+          JSG_REQUIRE(list.size() == 0, Error, "Transfer list is not supported");
+        }
+      }
+    }
+  }
+
+  KJ_IF_SOME(o, other) {
+    // TODO(message-port): Set up the external handler to support more types.
+    jsg::Serializer ser(js);
+    KJ_IF_SOME(d, data) {
+      ser.write(js, d.getHandle(js));
+    }
+    auto released = ser.release();
+    JSG_REQUIRE(released.sharedArrayBuffers.size() == 0, Error, "SharedArrayBuffer is unsupported");
+
+    capnp::MallocMessageBuilder builder;
+    rpc::JsValue::Builder val = builder.initRoot<rpc::JsValue>();
+    val.setV8Serialized(kj::mv(released.data));
+    o->deliverMessage(js, val.asReader());
+  }
+}
+
+// Should close this port, the entangle port, and any known rpc clients.
+void MessagePort::close() {
+  state = Closed{};
+  rpcClients.clear();
+  KJ_IF_SOME(o, other) {
+    auto closing = kj::mv(o);
+    other = kj::none;
+    closing->close();
+  }
+}
+
+// Start delivering messages on this port. Any messages that are
+// buffered will be drained immediately.
+void MessagePort::start(jsg::Lock& js) {
+  KJ_SWITCH_ONEOF(state) {
+    KJ_CASE_ONEOF(pending, Pending) {
+      auto list = kj::mv(pending);
+      state = Started{};
+      for (auto& item: list) {
+        ioContext.addTask(sendToRpc(capnp::clone(*item)));
+        // Local delivery
+        deliver(js, kj::mv(item), addRef());
+      }
+    }
+    KJ_CASE_ONEOF(_, Started) {
+      // Nothing to do in this case. We are already started!
+    }
+    KJ_CASE_ONEOF(_, Closed) {
+      // Nothing to do in this case. Can't start after closing.
+    }
+  }
+}
+
+kj::Maybe<jsg::JsValue> MessagePort::getOnMessage(jsg::Lock& js) {
+  return onmessageValue.map(
+      [&](jsg::JsRef<jsg::JsValue>& ref) -> jsg::JsValue { return ref.getHandle(js); });
+}
+
+void MessagePort::setOnMessage(jsg::Lock& js, jsg::JsValue value) {
+  if (!value.isObject() && !value.isFunction()) {
+    onmessageValue = kj::none;
+    // If we have no handlers and no onmessage ...
+    if (getHandlerCount("message"_kj) == 0 && onmessageValue == kj::none) {
+      // ...Put the port back into a pending state where messages
+      // will be enqueued until another listener is attached.
+      state = Pending();
+    }
+  } else {
+    onmessageValue = jsg::JsRef<jsg::JsValue>(js, value);
+    start(js);
+  }
+}
+
+namespace {
+// The jsrpc handler that receives messages posted from the remote and
+// delivers them to the local port.
+class JsMessagePortImpl final: public rpc::JsMessagePort::Server {
+ public:
+  JsMessagePortImpl(IoContext& ctx, jsg::Ref<MessagePort> port)
+      : port(kj::mv(port)),
+        weakIoContext(ctx.getWeakRef()) {}
+  ~JsMessagePortImpl() {
+    port->close();
+  }
+
+  kj::Promise<void> call(CallContext context) override {
+    IoContext& ctx = JSG_REQUIRE_NONNULL(weakIoContext->tryGet(), Error,
```

Comment:
I think this should be a DOMException

Reply:
This is following the pattern used elsewhere for other types.